cpp/src/arrow/array/builder_nested.cc

```diff
@@ -99,6 +99,11 @@ Status ListBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
     RETURN_NOT_OK(value_builder_->FinishInternal(&items));
   }
 
+  // If the type hasn't been specified in the constructor, try to infer it
+  if (!std::dynamic_pointer_cast<ListType>(type_)->value_type()) {
```

Comment:
We use `arrow::internal::checked_cast` instead of `dynamic_cast` in most places. We could also make a `_pointer_` variant of that.

Reply:
Thanks, fixed!